### ISSUE TITLE
Dealing with redis timeouts (lettuce client only)?

### ISSUE THREAD

**Issue by bnutzer** [bucket4j/bucket4j#406]
Hi,

we are using Bucket4j for rate limiting of incoming REST requests. For distribution, we chose redis, and the `LettuceBasedProxyManager`. The redis URL used for setting up the RedisClient includes a "timeout" specification, which is intended for both connection and subsequent non-blocking commands.

However, this timeout affects synchronous connections only. The bucket4j implementations (for lettuce, and, as far as I can tell, for redisson) use asynchronous connections. When the redis connection breaks, this results in unlimited waiting for a token get operation. While this might be the correct behavior for some, it is problematic in our scenario, where the initial REST call will not be responded to.

An "outer" layer of "wait for token, and handle timeout" would be a last resort, but I think this case should be handled by the `LettuceBasedProxyManager`. The client's default timeout could be added to the manager, and the `RedisFuture.get()` in method `getFutureValue` could use that timeout. Another option would be a dedicated timeout, configured with a `withTimeout(...)` in the builder.

What do you think? Would you consider a pull request implementing the "withTimeout" option (only for lettuce)?

Thanks a lot!

PS: There's still a method argument called "redissonFuture" in the `LettuceBasedProxyManager.convertToCompletableFuture` :)

**Comment by vladimir-bukhtoyarov** [comment]
>  The redis URL used for setting up the RedisClient includes a "timeout" specification, which is intended for both connection and subsequent non-blocking commands. However, this timeout affects synchronous connections only.

It is very bad news.

Bucket4j provides some hack for such unfriendly libraries called "ExecutionStrategy".
```java
import io.github.bucket4j.distributed.proxy.ExecutionStrategy
...
Duration timeout = Duration.ofSeconds(5);
int threadCount = 20; // choose rigth thread count according your use cases
Executor executor = Executors.newFixedThreadPool(threadCount);
ExecutionStrategy executionStrategy = ExecutionStrategy.backgroundTimeBounded(executor, timeout);
ClientSideConfig clientSideConfig = ClientSideConfig.getDefault().withExecutionStrategy(executionStrategy);

LettuceBasedProxyManager proxyManager = LettuceBasedProxyManager.builderFor(redisClient)
        .withClientSideConfig(clientSideConfig)
        ...
        .build();
```

> What do you think? Would you consider a pull request implementing the "withTimeout" option (only for lettuce)?

Looks that ExecutionStrategy allows to achive the same goal + provides more strong guaranties

**Comment by bnutzer** [comment]
Hi,

thanks for sharing your ideas on this. I've played around with the code you provided -- works well and does the trick. I'll use this.

Creating another thread, however, is not a cheap operation, and in itself might result in problems such as exhausted thread pools. Directly passing a timout to the `RedisFuture.get()` reduces these risks.

[timeout-for-lettuce.patch](https://github.com/bucket4j/bucket4j/files/12609825/timeout-for-lettuce.patch)

Thanks a lot!

**Comment by vladimir-bukhtoyarov** [comment]
@bnutzer 

Thank you for patch. I will apply it in few weeks. But, in general I prefer to avoid adding features for dedicated integrations, it looks that timeout can be added for mostly integrations, for example all JDBC integrations can support timeouts via ```preparedStatement.setQueryTimeout```. I am thinking about moving **timeout** setting to the CliendSideConfig and apply timeouts for all integrations where possible.

**Comment by vladimir-bukhtoyarov** [comment]
Hello @bnutzer 

I will release a fix this weekend.

**Comment by vladimir-bukhtoyarov** [comment]
Has been released with version `8.9.0`

Example of usage:
```java
LettuceBasedProxyManager.builderFor(...)
                    .withExpirationStrategy(...)
                    .withClientSideConfig(ClientSideConfig.getDefault().withRequestTimeout(Duration.ofSeconds(3)))
                    .build()
````
